### PR TITLE
fix(linter): pre-filter ignore pattern

### LINKED ISSUE #2464: --ignore-pattern seems not working
```shell
➜  ~/Documents/fabric git:(main) ✗ nr lint:oxlint

> @modyqyw/fabric@10.3.9 lint:oxlint /Users/wurui/Documents/fabric
> oxlint --deny=correctness --deny=perf --ignore-pattern="./src/index.ts" --fix ./src/index.ts

Finished in 8ms on 1 file with 86 rules using 10 threads.
Found 0 warnings and 0 errors.
```

## Steps

1. `git clone git@github.com:ModyQyW/oxc-2464.git`
2. `pnpm install`
3. `pnpm run oxlint-without-wildcard`
4. `pnpm run oxlint-with-wildcard`

## Behaviors

For `pnpm run oxlint-without-wildcard`, I get `Finished in 12ms on 1 file with 203 rules using 10 threads.`, but I think it should be `on 0 files`.

```shell
➜  ~/Documents/oxc-2464 git:(main) pnpm run oxlint-without-wildcard            

> oxc-2464@1.0.0 oxlint-without-wildcard /Users/wurui/Documents/oxc-2464
> oxlint -D=all --ignore-pattern=a.js --fix a.js


  ⚠ eslint-plugin-unicorn(no-empty-file): Empty files are not allowed.
   ╭─[a.js:1:1]
   ╰────
  help: Delete this file or add some code to it.

Finished in 12ms on 1 file with 203 rules using 10 threads.
Found 1 warning and 0 errors.
```

For `pnpm run oxlint-with-wildcard`, I get `Finished in 10ms on 1 file with 203 rules using 10 threads.`, but I think it should be `on 0 files`.

```shell
➜  ~/Documents/oxc-2464 git:(main) pnpm run oxlint-with-wildcard 

> oxc-2464@1.0.0 oxlint-with-wildcard /Users/wurui/Documents/oxc-2464
> oxlint -D=all --ignore-pattern=**/*.js --fix src/a.js


  ⚠ eslint-plugin-unicorn(no-empty-file): Empty files are not allowed.
   ╭─[src/a.js:1:1]
   ╰────
  help: Delete this file or add some code to it.

Finished in 10ms on 1 file with 203 rules using 10 threads.
Found 1 warning and 0 errors.
```

## My env

```shell
➜  ~/Documents/oxc-2464 git:(main) npx envinfo --system --binaries --browsers
Need to install the following packages:
envinfo@7.11.1
Ok to proceed? (y) 

  System:
    OS: macOS 14.2.1
    CPU: (10) arm64 Apple M2 Pro
    Memory: 57.69 MB / 16.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.11.0 - ~/Library/Caches/fnm_multishells/4367_1708500473819/bin/node
    Yarn: 1.22.19 - ~/Library/Caches/fnm_multishells/4367_1708500473819/bin/yarn
    npm: 10.2.4 - ~/Library/Caches/fnm_multishells/4367_1708500473819/bin/npm
    pnpm: 8.6.0 - ~/Library/Caches/fnm_multishells/4367_1708500473819/bin/pnpm
    bun: 1.0.28 - /opt/homebrew/bin/bun
    Watchman: 2024.01.22.00 - /opt/homebrew/bin/watchman
  Browsers:
    Chrome: 121.0.6167.184
    Safari: 17.2.1

```

### PR DESCRIPTION
**Pull request by mysteryven**
fix: #2464

### CODE REVIEW

**Comment by mysteryven** [comment]
Current dependencies on/for this PR:
* **#2472** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/2472?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/oxc-project/oxc/2472?utm_source=stack-comment).

**Comment by Boshen** [comment]
There's probably something broken with adding ignore patterns to `walk`. We'll also need to add tests.

**Comment by mysteryven** [comment]
Still need some time :( if file's path passed to `Walk`, the `ignorePattern` becomes invalided.

```js
fn main() {
    let mut walker = WalkBuilder::new("./a/foo.js");
    let mut override_builder = OverrideBuilder::new(Path::new("/"));
    
	let pattern = format!("!foo.js");
    override_builder.add(&pattern);
    
	let overrides = override_builder.build().unwrap();
    walker.overrides(overrides);

    for result in walker.build() {
        match result {
            Ok(entry) => println!("{}", entry.path().display()),
            Err(err) => println!("ERROR: {}", err),
        }
    }
}
```

Output: 

```bash
/a/foo.js
```

```bash
tree a/
a/
└── foo.js

0 directories, 1 file
```

File: crates/oxc_cli/src/walk.rs [hunk 1]
```diff
@@ -134,19 +118,97 @@ impl Walk {
...
+        for pattern in patterns {
+            override_builder.add(pattern).unwrap();
+        }
+        let overrides = override_builder.build().unwrap();
+        walk_builder.overrides(overrides);
+    }
 }
 
-#[cfg(test)]
+#[cfg(all(test, not(target_os = "windows")))]
```
Comment (Boshen) [inline crates/oxc_cli/src/walk.rs:186]: What's breaking on windows?

**Comment by Boshen** [comment]
This is hard ... I'll take a look at this soon.

**Comment by Boshen** [comment]
See https://github.com/oxc-project/oxc/issues/2464#issuecomment-2067964000

**Pull request closed without merging**